TYPESCRIPT.md

```diff
@@ -0,0 +1,209 @@
+    <img src="https://data.jsdelivr.com/v1/package/npm/contentful/badge" alt="jsDelivr Hits">
+  </a>
+<a href="https://npm-stat.com/charts.html?package=contentful">
+    <img src="https://img.shields.io/npm/dm/contentful.svg" alt="NPM downloads">
+  </a>
+<a href="https://unpkg.com/contentful/dist/contentful.browser.min.js">
+    <img src="https://img.badgesize.io/https://unpkg.com/contentful/dist/contentful.browser.min.js?compression=gzip" alt="GZIP bundle size">
+  </a>
+
+With version `10.0.0` we have completely rewritten the client to give the user more support on types.
+
+
+
+## Query types
+When querying for entries and assets, you get full type support for keys and values.
+This applies to:
+- `getEntry`
+- `getEntries`
+- `getAsset`
+- `getAssets`
+
+We have 2 levels of support:
+
+### Static query keys
+![](images/static-query-keys.png)
+
+```
+getEntries({
+    'skip': 10,
+    'limit': 20,
+    'include': 5
+})
+```
+
+### Dynamic (field) query keys
+![](images/dynamic-query-keys.png)
+
+Dynamic query keys are based on the given shape of the expected entries content type.
+To calculate dynamic keys, we have to pass-in the shape of the fields:
+```typescript
+type ExampleEntryFields = {
+  productName: Contentful.EntryFields.Text,
+  image: Contentful.Asset,
+  price: Contentful.EntryFields.Number,
+  categories: Contentful.Entry<CategoryEntryFields>[],
+}
+```
+
+This gives us the relevant information needed, to calculate the dynamic keys and their possible value types.
+```typescript
+getEntries<ExampleEntryFields>({
+    'fields.price[in]': [10, 20]
+})
+```
+
+#### Limitation
+Due to the limitation with recursive types, we can only calculate keys on the root level of your content type.
+There is currently no way to calculate keys for nested (recursive) content types.
+
+## Response types
+With version `10.0.0` we introduce [chained clients](./README.md#chained-clients) to make better assumptions on response types. 
+
+
+### `withAllLocales`
+If the current chain includes `withAllLocales`, `getEntry` and `getEntries` expect an optional second generic parameter for all existing locales in your space.
+If the `Locale` type is provided, your response type will define all locale keys for your field values:
+
+```typescript
+import * as contentful from "contentful";
+
+const client = contentful.createClient({
+  space: "<space-id>",
+  accessToken: "<content-delivery-token>",
+});
+
+const Fields = { productName: Contentful.EntryFields.Text }
+const Locales = 'en-US' | 'de-DE';
+const entry = client.withAllLocales.getEntry<Fields, Locales>() 
+```
+
+The return type of `getEntry` is matching the shape
+```json
+{
+  "fields": {
+    "productName": {
+      "de-DE": "<field-value>",
+      "en-US": "<field-value>"
+    }
+  }
+}
+```
+
+### `withoutLinkResolution`
+If the current chain includes `withoutLinkResolution`, the returned type doesn't define linked entities as Link objects.
```

Comment:
Do you mean the opposite? With `withoutLinkResolution` linked entities stay Link objects (as opposed to being resolved)